### PR TITLE
[MM][CG] Support ViT CUDA Graph for Gemma-4

### PR DESCRIPTION
**Pull request by anthonsu**
This PR implements static CUDA graph support for the Gemma-4 vision encoder.

## Purpose
This PR introduces full `SupportsEncoderCudaGraph` support for `Gemma4ForConditionalGeneration`, enabling 100% static compilation for the vision encoder. The core optimization bypasses dynamic slicing in the pooler. Instead of relying on data-dependent control flow, the host calculates static pointwise mapping indices (`gather_indices`). Inside the compiled graph, we execute a flat, fixed-shape gather operation to safely extract and pack valid features into the static token budget slots.

## Test Plan
1. **End-to-end Numerical Equivalence & CUDA Graph validation:**
   ```bash
   pytest tests/models/multimodal/generation/test_vit_cudagraph.py -k "gemma4" -v
   ```
2. CPU Dry-Run & Shape Math Verification: A standalone script was used to mock tensor creation to verify that the calculated padding sizes, offsets, and gather_indices mathematically align with the static token budgets.

<details><summary><b>Click to expand: CPU Shape Verification Script</b></summary>

```python
import torch

from vllm.model_executor.models.gemma4_mm import Gemma4ForConditionalGeneration


def verify():
    # 1. Mock vision tower config & model configuration
    class DummyVisionConfig:
        patch_size = 14
        num_channels = 3
        hidden_size = 1152
        standardize = False
        pooling_kernel_size = 2
        num_soft_tokens = 280

    class DummyTextConfig:
        hidden_size = 3072
        max_position_embeddings = 8192

    class DummyModelConfig:
        text_config = DummyTextConfig()
        vision_config = DummyVisionConfig()

    config = DummyModelConfig()

    class MockGemma4:
        def __init__(self):
            self.config = config

            class DummyVisionTower:
                config = DummyVisionConfig()

            self.vision_tower = DummyVisionTower()
            self.vllm_config = None

        get_encoder_cudagraph_config = (
            Gemma4ForConditionalGeneration.get_encoder_cudagraph_config
        )
        get_encoder_cudagraph_budget_range = (
            Gemma4ForConditionalGeneration.get_encoder_cudagraph_budget_range
        )
        get_input_modality = Gemma4ForConditionalGeneration.get_input_modality
        get_encoder_cudagraph_item_specs = (
            Gemma4ForConditionalGeneration.get_encoder_cudagraph_item_specs
        )
        prepare_encoder_cudagraph_capture_inputs = (
            Gemma4ForConditionalGeneration.prepare_encoder_cudagraph_capture_inputs
        )
        prepare_encoder_cudagraph_replay_buffers = (
            Gemma4ForConditionalGeneration.prepare_encoder_cudagraph_replay_buffers
        )
        get_max_frames_per_video = (
            Gemma4ForConditionalGeneration.get_max_frames_per_video
        )

    model = MockGemma4()

    # 1. Test get_encoder_cudagraph_config
    print("Testing get_encoder_cudagraph_config...")
    cg_config = model.get_encoder_cudagraph_config()
    print("Modalities:", cg_config.modalities)
    print("Buffer keys:", cg_config.buffer_keys)
    assert "pixel_values" in cg_config.buffer_keys
    assert "gather_indices" in cg_config.buffer_keys

    # Test get_max_frames_per_video
    print("\nTesting get_max_frames_per_video...")
    max_frames = model.get_max_frames_per_video()
    print("Max frames per video:", max_frames)
    assert max_frames == 32

    # 2. Test get_encoder_cudagraph_budget_range
    print("\nTesting get_encoder_cudagraph_budget_range...")

    class MockSchedulerConfig:
        max_num_batched_tokens = 2048

    class MockModelConfig:
        max_model_len = 8192

    class MockVllmConfig:
        scheduler_config = MockSchedulerConfig()
        model_config = MockModelConfig()

    vllm_config = MockVllmConfig()
    model.vllm_config = vllm_config

    min_budget, max_budget = model.get_encoder_cudagraph_budget_range(vllm_config)
    print(f"Budget Range: min={min_budget}, max={max_budget}")
    assert min_budget == 70
    assert max_budget == 2048

    # 3. Test get_encoder_cudagraph_item_specs for image
    print("\nTesting get_encoder_cudagraph_item_specs...")
    dummy_pv = torch.randn(256, 588)
    dummy_pp = torch.zeros(256, 2, dtype=torch.long)
    mm_kwargs = {"pixel_values": [dummy_pv], "pixel_position_ids": [dummy_pp]}
    specs = model.get_encoder_cudagraph_item_specs(mm_kwargs)
    print(
        f"Image Item Specs: input_size={specs[0].input_size}, "
        f"output_tokens={specs[0].output_tokens}"
    )
    assert specs[0].input_size == 256
    assert specs[0].output_tokens == 64

    # 4. Test prepare_encoder_cudagraph_capture_inputs
    print("\nTesting prepare_encoder_cudagraph_capture_inputs...")
    token_budget = 256
    max_batch_size = 2
    max_frames_per_batch = 8
    capture_inputs = model.prepare_encoder_cudagraph_capture_inputs(
        token_budget=token_budget,
        max_batch_size=max_batch_size,
        max_frames_per_batch=max_frames_per_batch,
        device=torch.device("cpu"),
        dtype=torch.float32,
    )
    print("Capture input keys:", list(capture_inputs.values.keys()))
    pixel_values_shape = capture_inputs.values["pixel_values"].shape
    gather_indices_shape = capture_inputs.values["gather_indices"].shape
    print("pixel_values shape:", pixel_values_shape)
    print("gather_indices shape:", gather_indices_shape)

    # max_size = 8. per_item_output = min(256, 280) = 256.
    # per_item_patches = 256 * 4 = 1024.
    assert pixel_values_shape == (8, 1024, 588)
    assert gather_indices_shape == (256,)

    # 5. Test prepare_encoder_cudagraph_replay_buffers (image)
    print("\nTesting prepare_encoder_cudagraph_replay_buffers...")
    replay_buffers = model.prepare_encoder_cudagraph_replay_buffers(
        mm_kwargs=mm_kwargs,
        max_batch_size=max_batch_size,
        max_frames_per_batch=max_frames_per_batch,
    )
    print("Replay buffer keys:", list(replay_buffers.values.keys()))
    replay_pv_shape = replay_buffers.values["pixel_values"].shape
    replay_gather_shape = replay_buffers.values["gather_indices"].shape
    print("Replay pixel_values shape:", replay_pv_shape)
    print("Replay gather_indices shape:", replay_gather_shape)

    assert replay_pv_shape == (1, 256, 588)
    # Total tokens = 64. Replay gather_indices is unpadded (size 64).
    assert replay_gather_shape == (64,)

    # With total_tokens=64, per_item_output = 256 // 4 = 64.
    # Since i=0, src_start = 0. src_end = 64.
    expected_indices = torch.arange(0, 64, dtype=torch.long)
    assert torch.equal(replay_buffers.values["gather_indices"], expected_indices)
    print("gather_indices values match expected output layout!")

    print("\nALL CPU SHAPE AND MATH ASSERTIONS PASSED SUCCESSFULLY!")


if __name__ == "__main__":
    verify()

```
</details>

<details><summary><b>Click to expand: Gemma4 vllm serve </b></summary>

```shell
VLLM_USE_V2_MODEL_RUNNER=0 \
vllm serve google/gemma-4-31B-it \
    --max-model-len=3072 \
    --tensor-parallel-size 1 \
    --max-num-batched-tokens 8192 \
    --no-enable-prefix-caching \
    --async-scheduling \
    --limit-mm-per-prompt '{"image": 8, "video": 0, "audio": 0}' \
    --mm-encoder-attn-backend FLASHINFER \
    --compilation-config '{"cudagraph_mm_encoder": true, "encoder_cudagraph_max_vision_items_per_batch": 4}'
```

</details>

## Test Result
*  [x] Local CPU shape math verification and index mapping logic passed successfully.
  ```bash
  $ ~/verify_shapes_cpu.py
INFO 07-22 17:16:50 [importing.py:53] Triton is installed but 0 active driver(s) found (expected 1). Disabling Triton to prevent runtime errors.
INFO 07-22 17:16:50 [importing.py:88] Triton not installed or not compatible; certain GPU-related functions will not be available.
WARNING 07-22 17:16:50 [interface.py:368] Failed to import from vllm._C: ModuleNotFoundError("No module named 'vllm._C'")
WARNING 07-22 17:16:50 [interface.py:368] Failed to import from vllm._C: ModuleNotFoundError("No module named 'vllm._C'")
WARNING 07-22 17:16:50 [interface.py:368] Failed to import from vllm._C: ModuleNotFoundError("No module named 'vllm._C'")
WARNING 07-22 17:16:51 [interface.py:368] Failed to import from vllm._C: ModuleNotFoundError("No module named 'vllm._C'")
Testing get_encoder_cudagraph_config...
Modalities: ['image', 'video']
Buffer keys: ['pixel_values', 'pixel_position_ids', 'gather_indices']
Testing get_max_frames_per_video...
Max frames per video: 32
Testing get_encoder_cudagraph_budget_range...
Budget Range: min=70, max=2048
Testing get_encoder_cudagraph_item_specs...
Image Item Specs: input_size=256, output_tokens=64
Testing prepare_encoder_cudagraph_capture_inputs...
Capture input keys: ['pixel_values', 'pixel_position_ids', 'gather_indices']
pixel_values shape: torch.Size([8, 1024, 588])
gather_indices shape: torch.Size([256])
Testing prepare_encoder_cudagraph_replay_buffers...
Replay buffer keys: ['pixel_values', 'pixel_position_ids', 'gather_indices']
Replay pixel_values shape: torch.Size([1, 256, 588])
Replay gather_indices shape: torch.Size([64])
gather_indices values match expected output layout!
ALL CPU SHAPE AND MATH ASSERTIONS PASSED SUCCESSFULLY!

  ```
* [x] Code hygiene (linters and formatters) clean.
* [x] GPU (H100) end-to-end **numerical equivalence validation, throughput performance and accuracy validation**, cuda-graph vs. no cuda-graph against MMMU-Pro.
* [x] CI/CD pipeline & build passed.

<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/models/gemma4_mm.py [hunk 1]
```diff
@@ -1517,6 +1528,402 @@ def embed_multimodal(self, **kwargs: object) -> MultiModalEmbeddings:
...
+        dtype: torch.dtype,
+        path: str = "default",
+    ) -> "EncoderCudaGraphCaptureInputs":
+        from vllm.v1.worker.encoder_cudagraph_defs import (
+            EncoderCudaGraphCaptureInputs,
+        )
+
+        import math
+        max_size = max(max_batch_size, max_frames_per_batch)
+        # per_item_patches must satisfy k² × per_item_output == per_item_patches
```
Comment (lk-chen) [inline vllm/model_executor/models/gemma4_mm.py]: These logic is a bit off, plz check https://docs.vllm.ai/en/stable/design/cuda_graphs_multimodal/#dual-path-graph-capture:~:text=Greedily%20pack%20images%3A%20an%20image%20is%20added%20to%20the%20current%20sub%2Dbatch%20only%20if%20both%20the%20accumulated%20global%20tokens%20%E2%89%A4%20max_global_budget%20and%20the%20accumulated%20local%20tokens%20%E2%89%A4%20max_local_budget%2C%20with%20the%20image%20count%20%E2%89%A4%20max_batch_size for how padding and budget fit work

Comment (anthonsu) [reply vllm/model_executor/models/gemma4_mm.py]: Resolved the issue. The global token_budget is distributed to the local budget.

File: vllm/model_executor/models/gemma4_mm.py [hunk 2]
```diff
@@ -1522,6 +1532,396 @@ def embed_multimodal(self, **kwargs: object) -> MultiModalEmbeddings:
...
+                "pixel_values": pad_pixel_values,
+                "pixel_position_ids": pad_pixel_position_ids,
+            },
+        )
+
+    def get_encoder_cudagraph_budget_range(
+        self,
+        vllm_config: VllmConfig,
+    ) -> tuple[int, int]:
+        min_budget = 64
```
Comment (kwang3939) [inline vllm/model_executor/models/gemma4_mm.py]: The smallest budget is `70`? https://github.com/anthonsu/vllm/blob/fc28984c1d9965600bcaafc00e185d2de2c8f42c/vllm/model_executor/models/gemma4_mm.py#L97

File: vllm/model_executor/models/gemma4_mm.py [hunk 3]
```diff
@@ -1522,6 +1532,396 @@ def embed_multimodal(self, **kwargs: object) -> MultiModalEmbeddings:
...
+        max_budget = min(
+            vllm_config.scheduler_config.max_num_batched_tokens,
+            self.config.text_config.max_position_embeddings,
```
Comment (kwang3939) [inline vllm/model_executor/models/gemma4_mm.py]: To match with other models, I think this should be `self.model_config.max_model_len,`

File: vllm/model_executor/models/gemma4_mm.py [hunk 4]
```diff
@@ -1522,6 +1532,396 @@ def embed_multimodal(self, **kwargs: object) -> MultiModalEmbeddings:
...
+        from vllm.v1.worker.encoder_cudagraph_defs import EncoderItemSpec
+
+        modality = self.get_input_modality(mm_kwargs)
+        if modality == "image":
+            pixel_values = mm_kwargs["pixel_values"]
+            if isinstance(pixel_values, list):
+                return [
+                    EncoderItemSpec(
+                        input_size=pv.shape[0],
+                        output_tokens=pv.shape[0] // 4,
```
Comment (kwang3939) [inline vllm/model_executor/models/gemma4_mm.py]: Maybe use `vision_cfg.pooling_kernel_size**2` instead? Also I think it's `9` instead of `4` for Gemma 4.

File: vllm/model_executor/models/gemma4_mm.py (hunk 2 above)
Comment (anthonsu) [reply vllm/model_executor/models/gemma4_mm.py]: Yes, I will make the min budget match this SOFT_TOKENS setting.

File: vllm/model_executor/models/gemma4_mm.py [hunk 5]
```diff
@@ -1522,6 +1532,396 @@ def embed_multimodal(self, **kwargs: object) -> MultiModalEmbeddings:
...
+
+        max_size = max(max_batch_size, max_frames_per_batch)
+        per_item_output = (token_budget + max_size - 1) // max_size
+
+        if modality == "image":
+            dst_offset = 0
+            for i, n_tok in enumerate(per_item_out_tokens):
+                src_start = i * per_item_output
+                src_end = src_start + n_tok
+                gather_indices[dst_offset : dst_offset + n_tok] = torch.arange(
```
[the remark this answers is not in the document]
Comment (anthonsu) [reply vllm/model_executor/models/gemma4_mm.py]: Fixed.

**Comment by lk-chen** [comment]
tested on H100 with
```bash
VLLM_USE_V2_MODEL_RUNNER=0 FLASHINFER_DISABLE_VERSION_CHECK=1 \
    vllm serve google/gemma-4-31B-it \
      --max-model-len=3072 --tensor-parallel-size 1 --max-num-batched-tokens 8192 \
      --no-enable-prefix-caching --async-scheduling \
      --limit-mm-per-prompt '{"image": 8, "video": 0, "audio": 0}' \
      --mm-encoder-attn-backend FLASHINFER \
      --compilation-config '{"cudagraph_mm_encoder": true, "encoder_cudagraph_token_budgets": [64, 128, 256, 512, 1024],
  "encoder_cudagraph_max_vision_items_per_batch": 8, "encoder_cudagraph_max_frames_per_batch": 4}
```

File: tests/models/multimodal/generation/test_vit_cudagraph.py [hunk 6]
```diff
@@ -249,6 +249,26 @@ def gemma3_chat_template(content: str) -> str:
         },
         skip=True,  # TODO: Re-enable this once OOM issues are resolved on CI.
     ),
+    "gemma4": VitCudagraphTestConfig(
+        model="google/gemma-4-31B-it",
```
Comment (Isotr0py) [inline tests/models/multimodal/generation/test_vit_cudagraph.py]:
```suggestion
        model="google/gemma-4-E2B-it",
```
We can use 2B model here, then we don't need to use the dummy ones.

**Review by Isotr0py — APPROVED** [review]
Overall look reasonable, thanks!

File: tests/models/multimodal/generation/test_vit_cudagraph.py [hunk 7]
```diff
@@ -250,7 +250,7 @@ def gemma3_chat_template(content: str) -> str:
         skip=True,  # TODO: Re-enable this once OOM issues are resolved on CI.
     ),
     "gemma4": VitCudagraphTestConfig(
-        model="google/gemma-4-31B-it",
+        model="google/gemma-4-E2B-it",
```
Comment (AndreasKaratzas) [inline tests/models/multimodal/generation/test_vit_cudagraph.py:253]: I am in favor of using a smaller model, but why was that change necessary? Was the current model failing?

Comment (anthonsu) [reply tests/models/multimodal/generation/test_vit_cudagraph.py:253]: No, the previous build did not fail under the 31B model with dummy weights. 

This switch was made following @Isotr0py's suggestion above:

"_We can use 2B model here, then we don't need to use the dummy ones._"

**Comment by anthonsu** [comment]
As requested by the Mergify check, I have successfully rebased this branch on top of the latest main branch. The workflow permission and the auto-rebase check should clear now.

@Isotr0py, @AndreasKaratzas, could you please help with merging this PR? (I don't seem have the write/merge permissions the repo.) Thank you!!
